File: GRAMATICAS/GRAMATICA3/gramatica3.py
def acepta_cadena(cadena):
    n = 0
    length = len(cadena)

    # Contar 'a'
    while n < length and cadena[n] == 'a':
        n += 1

    # Verificar que el resto sean solo 'b'
    for i in range(n, length):
        if cadena[i] != 'b':
            return False

    # Condición: n > 0 y cantidad de b's = n + 1
    return n > 0 and (length - n) == (n + 1)

def main():
    archivo = input("Ingrese el nombre del archivo: ")

    try:
        with open(archivo, "r") as f:
            for linea in f:
                cadena = linea.strip()
                if acepta_cadena(cadena):
                    print("acepta")
                else:
                    print("no acepta")
    except FileNotFoundError:
        print("No se pudo abrir el archivo")

if __name__ == "__main__":
    main()

File: GRAMATICAS/GRAMATICA3/gramatica3.c
#include <stdio.h>
#include <string.h>

int acepta_cadena(const char *cadena) {
    int n = 0;
    int len = strlen(cadena);

    // Contar 'a'
    while (n < len && cadena[n] == 'a') {
        n++;
    }
    // El resto deben ser 'b'
    for (int i = n; i < len; i++) {
        if (cadena[i] != 'b') return 0;
    }
    // Condición: n > 0 y cantidad de b's == n + 1
    return (n > 0) && ((len - n) == (n + 1));
}
int main() {
    char archivo[100], linea[100];
    FILE *fp;

    printf("Ingrese el nombre del archivo: ");
    scanf("%s", archivo);

    fp = fopen(archivo, "r");
    if (!fp) {
        perror("No se pudo abrir el archivo");
        return 1;
    }

    while (fgets(linea, sizeof(linea), fp)) {
        // Eliminar salto de línea
        linea[strcspn(linea, "\n")] = '\0';

        if (acepta_cadena(linea)) {
            printf("acepta\n");
        } else {
            printf("no acepta\n");
        }
    }

    fclose(fp);
    return 0;
}
